stan/math/opencl/rev/vari.hpp

```diff
@@ -0,0 +1,209 @@
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+ protected:
+  // to allow access to this constructor from instantinations with different
+  // template parameters
+  template <typename, typename>
+  friend class vari_value;
+  /**
+   * Construct a matrix_cl variable implementation from a value and
+   * adjoint.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @param val Value of the constructed variable.
+   * @param adj Adjoint of the constructed variable.
+   */
+  vari_value(T&& val, T&& adj)
+      : chainable_alloc(),
+        adj_(std::forward<T>(adj)),
+        val_(std::forward<T>(val)) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+ public:
+  /**
+   * Construct an matrix_cl variable implementation from a value. The
+   *  adjoint is initialized to zero and if `stacked` is `false` this vari
+   *  will be not be put on the var_stack. Instead it will only be put on
+   *  a stack to keep track of whether the adjoint needs to be set to zero.
+   *
+   * All constructed variables are added to a stack.  Variables
+   *  should be constructed before variables on which they depend
+   *  to insure proper partial derivative propagation.  During
+   *  derivative propagation, the chain() method of each variable
+   *  will be called in the reverse order of construction.
+   *
+   * @tparam S A `matrix_cl` or kernel generator expression type that is
+   * convertible to `value_type`
+   * @param x Value of the constructed variable.
+   * @param stacked If false will put this this vari on the nochain stack so
+   * that its `chain()` method is not called.
+   */
+  template <typename S, require_convertible_t<S&, T>* = nullptr>
+  vari_value(S&& x, bool stacked)
+      : chainable_alloc(),
+        adj_(constant(0, x.rows(), x.cols())),
+        val_(std::forward<S>(x)) {
+    if (stacked) {
+      ChainableStack::instance_->var_stack_.push_back(this);
+    } else {
+      ChainableStack::instance_->var_nochain_stack_.push_back(this);
+    }
+  }
+
+  /**
+   * Returns a view into a block of matrix.
+   * @param row starting row of the block
+   * @param col starting column of the block
+   * @param rows number of rows in the block
+   * @param cols number of columns in the block
+   * @return block
+   */
+  auto block(int row, int col, int rows, int cols) {
+    auto&& val_block = stan::math::block(val_, row, col, rows, cols);
+    auto&& adj_block = stan::math::block(adj_, row, col, rows, cols);
+    return vari_value<std::decay_t<decltype(val_block)>>(std::move(val_block),
+                                                         std::move(adj_block));
+  }
```

Comment:
[Q] This is just something that I'm not sure about

When we do views like blocks etc do they need to be put on the `var_stack_`? idt we will ever do `.chain()` on this vari will we? Since this makes a copy of the data we want to be able to zero it's adjoints, but I think that means this can just go on the `var_nochain_stack_`. So we could use something like 

```cpp
 // false to put on nochain stack
vari_value<std::decay_t<decltype(val_block)>>(std::move(val_block),
                                                         std::move(adj_block), false);
```

Reply:
Good point. However we also don't need to ever zero its adjoints. They will be zeroed when the matrix, this is view into, is zeroed.